File: server/nn/nn.go
package nn

// nn is a Neural Network interface layer

// Detection is an object that a neural network has found in an image
type Detection struct {
	Class      int     `json:"class"`
	Confidence float32 `json:"confidence"`
	Box        Rect    `json:"box"`
}

type DetectionResult struct {
	CameraID    int64       `json:"cameraID"`
	ImageWidth  int         `json:"imageWidth"`
	ImageHeight int         `json:"imageHeight"`
	Objects     []Detection `json:"objects"`
}

type DetectionParams struct {
	ProbabilityThreshold float32 // Value between 0 and 1. Lower values will find more objects. Zero value will use the default.
	NmsThreshold         float32 // Value between 0 and 1. Lower values will merge more objects together into one. Zero value will use the default.
}

func DefaultDetectionParams() *DetectionParams {
	return &DetectionParams{
		ProbabilityThreshold: 0.5,
		NmsThreshold:         0.45,
	}
}

// ObjectDetector is given an image, and returns zero or more detected objects
type ObjectDetector interface {
	// Close closes the detector (you MUST call this when finished, because it's a C++ object underneath)
	Close()
	// DetectObjects returns a list of objects detected in the image
	DetectObjects(nchan int, image []byte, width, height int, params *DetectionParams) ([]Detection, error)
}

const (
	COCOPerson     = 0
	COCOBicycle    = 1
	COCOCar        = 2
	COCOMotorcycle = 3
	COCOAirplane   = 4
	COCOBus        = 5
	COCOTrain      = 6
	COCOTruck      = 7
)

// COCO classes
var COCOClasses = []string{
	"person",
	"bicycle",
	"car",
	"motorcycle",
	"airplane",
	"bus",
	"train",
	"truck",
	"boat",
	"traffic light",
	"fire hydrant",
	"stop sign",
	"parking meter",
	"bench",
	"bird",
	"cat",
	"dog",
	"horse",
	"sheep",
	"cow",
	"elephant",
	"bear",
	"zebra",
	"giraffe",
	"backpack",
	"umbrella",
	"handbag",
	"tie",
	"suitcase",
	"frisbee",
	"skis",
	"snowboard",
	"sports ball",
	"kite",
	"baseball bat",
	"baseball glove",
	"skateboard",
	"surfboard",
	"tennis racket",
	"bottle",
	"wine glass",
	"cup",
	"fork",
	"knife",
	"spoon",
	"bowl",
	"banana",
	"apple",
	"sandwich",
	"orange",
	"broccoli",
	"carrot",
	"hot dog",
	"pizza",
	"donut",
	"cake",
	"chair",
	"couch",
	"potted plant",
	"bed",
	"dining table",
	"toilet",
	"tv",
	"laptop",
	"mouse",
	"remote",
	"keyboard",
	"cell phone",
	"microwave",
	"oven",
	"toaster",
	"sink",
	"refrigerator",
	"book",
	"clock",
	"vase",
	"scissors",
	"teddy bear",
	"hair drier",
	"toothbrush",
}

File: www/src/camera/nn.ts
export class Rect {
	x = 0;
	y = 0;
	width = 0;
	height = 0;

	parseJSON(j: any) {
		this.x = j.x;
		this.y = j.y;
		this.width = j.width;
		this.height = j.height;
	}
}

export class Detection {
	class = 0;
	confidence = 0;
	box = new Rect();
}

// Neural network detection result on a frame
export class DetectionResult {
	cameraID = 0;
	objects: Detection[] = [];
	imageWidth = 1;
	imageHeight = 1;

	static fromJSON(j: any): DetectionResult {
		let dr = new DetectionResult();
		dr.cameraID = j.cameraID;
		dr.imageWidth = j.imageWidth;
		dr.imageHeight = j.imageHeight;

		for (let jo of j.objects) {
			let d = new Detection();
			d.class = jo.class;
			d.confidence = jo.confidence;
			d.box.parseJSON(jo.box);
			dr.objects.push(d);
		}
		return dr;
	}
}

// An object that was detected by the Object Detector, and is now being tracked by a post-process
// SYNC-TRACKED-OBJECT
export class TrackedObject {
	id = 0;
	class = 0;
	box = new Rect();
	genuine = false;

	parseJSON(j: any) {
		this.id = j.id;
		this.class = j.class;
		this.box.parseJSON(j.box);
		this.genuine = j.genuine;
	}
}

// Analysis that runs as a post-process on top of the raw Object Detection neural network
// SYNC-ANALYSIS-STATE
export class AnalysisState {
	cameraID = 0;
	input: DetectionResult | null = null;
	objects: TrackedObject[] = [];

	static fromJSON(j: any): AnalysisState {
		let as = new AnalysisState();
		as.cameraID = j.cameraID;
		as.input = j.input ? DetectionResult.fromJSON(j.input) : null;
		for (let jo of j.objects) {
			let to = new TrackedObject();
			to.parseJSON(jo);
			as.objects.push(to);
		}
		return as;
	}
}
